bluetooth: l2cap: support sending buffers that are larger than MTU

mjchen0 opened the pull request:
If the net_buf passed to bt_l2cap_chan_send() is larger than the TX MTU, the l2cap layer will now send as many SDUs as needed to send the net_buf instead of returning an error.

The processing is similar to how a SDU that is larger than a PDU is fragmented into multiple PDUs. The net_buf is left on the tx_queue until it is done being sent.
Adding of the SDU header is moved from bt_l2cap_dyn_chan_send() to l2cap_data_pull(), just like how the PDU header is added.

Thalley reviewed (commented):
It's an interesting change request. Can you elaborate on the use case?

Splitting SDUs into multiple SDUs isn't something I've seen in the core spec. Is it defined? 

Imagine sending an L2CAP SDU with a header and some data - If the receive is always expecting the first e.g. 8 octets to be a header, then splitting the SDU into multiple would tell the transmitting Zephyr application that everything went well, while the 2 SDUs may be discarded by the receive because of the first SDU perhaps failing a length check, and the second one not having the header.

mjchen0 commented:
> It's an interesting change request. Can you elaborate on the use case?
> 
> Splitting SDUs into multiple SDUs isn't something I've seen in the core spec. Is it defined?
> 
> Imagine sending an L2CAP SDU with a header and some data - If the receive is always expecting the first e.g. 8 octets to be a header, then splitting the SDU into multiple would tell the transmitting Zephyr application that everything went well, while the 2 SDUs may be discarded by the receive because of the first SDU perhaps failing a length check, and the second one not having the header.

I'm not suggesting to suggest one SDU into multiple SDUs. I'm trying to support a use case where a single net_buf might have be so large that it's bigger than the MTU, so can't fit in a single SDU. The current API would just reject this net_buf with an error if it was passed to bt_l2cap_chan_send(), forcing the caller to figure out the best way to divide their large message into multiple SDUs themselves and then pass those to bt_l2cap_chan_send().

I thought it was cleaner and easier to just have this feature instead in l2cap.c, so it will go ahead and send the large net_buf in as many SDUs as needed. Unless I'm missing something where the bt_l2cap_chan_send() has to be limited to a net_buf that is less than the MTU size.

Thalley commented:
> I'm not suggesting to suggest one SDU into multiple SDUs. I'm trying to support a use case where a single net_buf might have be so large that it's bigger than the MTU, so can't fit in a single SDU. The current API would just reject this net_buf with an error if it was passed to bt_l2cap_chan_send(), forcing the caller to figure out the best way to divide their large message into multiple SDUs themselves and then pass those to bt_l2cap_chan_send().

That is a fair point. The current documentation does not explicitly state, but does imply (given references to `BT_L2CAP_SDU_CHAN_SEND_RESERVE`, `BT_L2CAP_SDU_BUF_SIZE`, etc.), that `bt_l2cap_chan_send` sends a single SDU, and your suggested modification would change this to implicitly split (too) large data, instead of providing an error to the caller. In those cases, you would be splitting a would-be single SDU into multiple SDUs. 

This is definitely a stable API change, and arguably a breaking stable API change, since any application that may have handled the return value of `EMSGSIZE` would now not get that. 

My gut feeling is that this needs to go through the steps of https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#introducing-breaking-api-changes, but I'll defer to the host maintainer, @alwa-nordic for that. Also keep in mind that this is just my input for now, and I'm not rejecting this change, but I think you need to provide a better reason than "just in case". 
An alternative could be to introduce a new `bt_l2cap_send_many` or something to avoid modifying existing APIs, but not sure if that's actually better.

mjchen0 committed:
bluetooth: l2cap: support sending buffers that are larger than MTU

If the net_buf passed to bt_l2cap_chan_send() is larger
than the TX MTU, the l2cap layer will now send as many
SDUs as needed to send the net_buf instead of returning
an error.

The processing is similar to how a SDU that is larger than
the MPS is fragmented into multiple PDUs. The net_buf is
left on the tx_queue until it is done being sent.
Adding of the SDU header is moved from bt_l2cap_dyn_chan_send()
to l2cap_data_pull(), just like how the PDU header is added.

Signed-off-by: Mike J. Chen <mjchen@google.com>

mjchen0 commented:
@alwa-nordic , I was busy for a few weeks but I have some time now to try to implement your suggestion for a seg_send callback, mirroring the seg_recv callback. A few questions:

1) do you think the host should manage the SDU header, or have the callback handle that? I'm not sure if the host even needs to know if the segment/PDU it's been given is the first of an SDU or not. If it's the one responsible for adding the SDU header and doing MTU checks, then it needs to know it is the first segment/PDU. Otherwise, it could just add PDU headers, check MPS and credits, and not worry about anything else.

2) how do you think app should initiate the transfer? A new API, or maybe call the bt_l2cap_chan_send() API with a NULL buf?

alwa-nordic commented:
> I was busy for a few weeks but I have some time now to try to implement your suggestion for a seg_send callback, mirroring the seg_recv callback. A few questions:
> 
>     1. do you think the host should manage the SDU header, or have the callback handle that? I'm not sure if the host even needs to know if the segment/PDU it's been given is the first of an SDU or not. If it's the one responsible for adding the SDU header and doing MTU checks, then it needs to know it is the first segment/PDU. Otherwise, it could just add PDU headers, check MPS and credits, and not worry about anything else.

My preference is that the callback fills out the SDU header. The Host should look at the header to validate it for the benefit of out users, but this could be optionally disabled. The validation logs an error if the SDU size exceeds MTU and also keeps track of the SDU length remaining and logs an errors if a PDU contains trailing data.

>     2. how do you think app should initiate the transfer? A new API, or maybe call the bt_l2cap_chan_send() API with a NULL buf?

I prefer a new API.
